storage/allocator.go

```diff
@@ -13,75 +13,289 @@
+}
+
+// capacityGossipUpdate is a gossip callback triggered whenever capacity
+// information is gossiped. It just tracks keys used for capacity
+// gossip.
+func (a *allocator) capacityGossipUpdate(key string, _ bool) {
+	a.Lock()
+	defer a.Unlock()
+
+	// Clear the cached store lists on new gossip.
+	a.storeLists = nil
+	if a.capacityKeys == nil {
+		a.capacityKeys = map[string]struct{}{}
+	}
+	a.capacityKeys[key] = struct{}{}
+}
 
-	// Randomly pick a node weighted by capacity.
-	var candidates []*proto.StoreDescriptor
-	var capacityTotal float64
+// AllocateTarget returns a suitable store for a new allocation with the
+// required attributes. Nodes already accommodating existing replicas
+// are ruled out as targets.
+func (a *allocator) AllocateTarget(required proto.Attributes, existing []proto.Replica) (
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	stores, err := a.selectRandom(3, required, existing, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Choose the store with the highest percent capacity available.
+	var leastStore *proto.StoreDescriptor
 	for _, s := range stores {
-		if _, ok := usedNodes[s.Node.NodeID]; !ok {
-			candidates = append(candidates, s)
-			capacityTotal += s.Capacity.PercentAvail()
+		if leastStore == nil || s.Capacity.FractionUsed() < leastStore.Capacity.FractionUsed() {
+			leastStore = s
+		}
+	}
+	return leastStore, nil
+}
+
+// RebalanceTarget returns a suitable store for a rebalance target
+// with required attributes. Rebalance targets are selected at random
+// from amongst stores which fall within an acceptable multiple of
+// standard deviations from either the used or range count means.
+func (a *allocator) RebalanceTarget(required proto.Attributes, existing []proto.Replica) (
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	filter := func(s *proto.StoreDescriptor, count, used *stat) bool {
+		// Use counts instead of capacities if the cluster has mean
+		// fraction used below a threshold level. This is primarily useful
+		// for balancing load evenly in nascent deployments.
+		if used.mean < minFractionUsedThreshold {
+			return float64(s.Capacity.RangeCount) < count.mean
+		}
+		maxFractionUsed := used.mean * (1 - rebalanceFromMean)
+		if maxFractionUsedThreshold < maxFractionUsed {
+			maxFractionUsed = maxFractionUsedThreshold
+		}
+		return s.Capacity.FractionUsed() < maxFractionUsed
+	}
+	stores, err := a.selectRandom(1, required, existing, filter)
+	if err != nil {
+		return nil, err
+	}
+	return stores[0], nil
+}
+
+// ShouldRebalance returns whether the specified store is overweight
+// according to the cluster mean and should rebalance a range.
+func (a *allocator) ShouldRebalance(s *proto.StoreDescriptor) bool {
+	a.Lock()
+	defer a.Unlock()
+	sl := a.getStoreList(*s.CombinedAttrs())
+
+	if sl.used.mean < minFractionUsedThreshold {
+		return float64(s.Capacity.RangeCount) > sl.count.mean
+	}
+	return s.Capacity.FractionUsed() > sl.used.mean
```

Comment:
This should have some margin for error (maybe add rebalanceFromMean, just like we subtract it when looking for targets) instead of trying to rebalance every range that is above the mean. The threshold should probably be derived from the number of ranges per store - when there are few ranges, moving one of them could cause a swing of more than 2.5%.